renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -157,28 +335,129 @@ void QueryPostModPixelValues(WrappedOpenGL *driver, GLPixelHistoryResources &res
+  }
+
+  for(size_t i = 0; i < modEvents.size(); i++)
+  {
     ModificationValue modValue;
-    PixelValue pixelValue;
-    driver->glReadPixels(x, y, 1, 1, eGL_RGBA, eGL_FLOAT, (void *)pixelValue.floatValue.data());
-    driver->glReadPixels(x, y, 1, 1, eGL_DEPTH_COMPONENT, eGL_FLOAT, (void *)&modValue.depth);
-    driver->glReadPixels(x, y, 1, 1, eGL_STENCIL_INDEX, eGL_INT, (void *)&modValue.stencil);
-    modValue.col = pixelValue;
+
+    for(int j = 0; j < 4; ++j)
+    {
+      modValue.col.floatValue[j] = colourValues[i * 4 + j];
+    }
+    modValue.depth = depthValues[i];
+    modValue.stencil = stencilValues[i];
 
     history[i].postMod = modValue;
+  }
+}
+
+// This function a) calculates the number of fagments per event
+//           and b) calculates the shader output values per event
+std::map<uint32_t, uint32_t> QueryNumFragmentsByEvent(WrappedOpenGL *driver,
+                                                      GLPixelHistoryResources &resources,
+                                                      const rdcarray<EventUsage> &modEvents,
+                                                      rdcarray<PixelModification> &history, int x,
+                                                      int y)
+{
+  driver->ReplayLog(0, modEvents[0].eventId, eReplay_WithoutDraw);
+
+  std::map<uint32_t, uint32_t> eventFragments;
+  rdcarray<float> colourValues;
+  colourValues.resize(4 * modEvents.size());
+  rdcarray<float> depthValues;
+  depthValues.resize(modEvents.size());
+  rdcarray<int> stencilValues;
+  stencilValues.resize(modEvents.size());
+
+  for(size_t i = 0; i < modEvents.size(); ++i)
+  {
+    driver->glStencilOp(eGL_INCR, eGL_INCR, eGL_INCR);
+    driver->glStencilMask(0xff);    // default for 1 byte
+    driver->glStencilFunc(eGL_ALWAYS, 0, 0xff);
+    driver->glClearStencil(0);
+    driver->glClear(eGL_STENCIL_BUFFER_BIT);
+    driver->glEnable(eGL_STENCIL_TEST);
+    // depth test enable
+    driver->glEnable(eGL_DEPTH_TEST);
+    driver->glDepthFunc(eGL_ALWAYS);
+    driver->glDepthMask(GL_TRUE);
+    driver->glDisable(eGL_BLEND);
+
+    // replay event
+    driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_OnlyDraw);
+
+    // Blit the values into out framebuffer
+    GLint savedReadFramebuffer, savedDrawFramebuffer;
+    driver->glGetIntegerv(eGL_DRAW_FRAMEBUFFER_BINDING, &savedDrawFramebuffer);
+    driver->glGetIntegerv(eGL_READ_FRAMEBUFFER_BINDING, &savedReadFramebuffer);
+    driver->glBindFramebuffer(eGL_DRAW_FRAMEBUFFER, resources.frameBuffer);
+    driver->glBindFramebuffer(eGL_READ_FRAMEBUFFER, savedDrawFramebuffer);
+
+    SafeBlitFramebuffer(
+        x, y, x + 1, y + 1, GLint(i), 0, GLint(i) + 1, 1,
+        eGL_COLOR_BUFFER_BIT |
+            (resources.depthTextureAttachedToFrameBuffer ? eGL_DEPTH_BUFFER_BIT : eGL_NONE) |
+            (resources.stencilTextureAttachedToFrameBuffer ? eGL_STENCIL_BUFFER_BIT : eGL_NONE),
+        eGL_NEAREST);
```

Comment:
This doesn't get the shader output values because it's reading from the target framebuffer, meaning the output has already been quantised by whatever format there is there. To get the shader output values we need to render offscreen to RGBA32F to be sure we get the correct full precision values.